File: 0000-0999/0222.CountCompleteTreeNodes.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func countNodes(root *TreeNode) int {
    if root ==  nil {
        return 0
    }
    // O(logN*logN): logN + (logN)-1 + (logN)-2 + ... + 1 = (1+logN)*logN/2
    hl, hr := 0, 0
    for p := root; p != nil; p = p.Left {
        hl++
    }
    for p := root; p != nil; p = p.Right {
        hr++
    }
    if hl == hr {
        return 2<<uint(hl-1) - 1
    }
    return 1 + countNodes(root.Left) + countNodes(root.Right)
    // O(N)
    //return 1 + countNodes(root.Left) + countNodes(root.Right)
}

File: 0000-0999/0222.CountCompleteTreeNodes.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def countNodes(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0
        l = self.height(root.left)
        r = self.height(root.right)
        if l == r:
            return 2**l + self.countNodes(root.right)
        else:
            return 2**r + self.countNodes(root.left)

    def height(self, node):
        if not node:
            return 0
        return 1 + self.height(node.left)
